llvm/lib/Target/DirectX/DXIL.td

```diff
@@ -738,6 +739,16 @@ def UMin : DXILOp<40, binary> {
   let attributes = [Attributes<DXIL1_0, [ReadNone]>];
 }
 
+def UAddc : DXILOp<44, binaryWithCarryOrBorrow > {
+  let Doc = "Unsigned 32-bit integer arithmetic add with carry. uaddc(a,b) = (a+b, a+b overflowed ? 1 : 0)";
+  let intrinsics = [IntrinSelect<int_uadd_with_overflow>];
```

Comment:
One thing I am a little worried about is if we are misusing `int_uadd_with_overflow` here. This kind feels like a square is a rectangle but rectangles are not squares cases where llvm is using `int_uadd_with_overflow` much more broadly than we are. It makes me think If something else were to come along and emit `int_uadd_with_overflow` and it wasn't a `uint32_t2` or `uint32_t4` case would we be doing the right thing by associating the overflow intrinsic with `UAddc`?

Reply:
It is the `AddUint64` HLSL function that only accepts `uint32_t2` and `uint32_t4` to perform a 64-bit add operation. `AddUint64` lowers into several instructions, one (or two) of which is the `int_uadd_with_overflow` intrinsic used to compute an i32 add with carry, which is the same exact operation performed by `UAddc`.

Reply:
`i16` and `i64` could be supported by `UAddc` as well with minimal effort, but I'm not sure if that would conform to spec. It would, however, still be equivalent to `int_uadd_with_overflow` behavior.

Reply:
We would need to rev the DXIL version to introduce UAddc for other types I think. I think Farzon's concern is that we might want to (at some point) handle `int_uadd_with_overflow` generically, and since there isn't a DXIL op for it that would presumably involve an expansion. If we use the 32 bit `int_uadd_with_overflow` to map to UAddc, then whatever pass did that expansion would need to know not to do it for certain overloads and that might be awkward.

Reply:
Whatever does the lowering for `int_uadd_with_overflow` generically would still need to know that the 32-bit `int_uadd_with_overflow` maps directly to the `UAddc` DXIL op in order to codegen optimally, as opposed to creating some expansion that does the same thing using more instructions.

Reply:
This isn't a nice tablegen solution, but we have the power to check  if we have a non 32 bit case in `DXILIntrinsicExpansion.cpp`. It would be a little bit of a new pattern for us because it would be the first time we check an intrinsic  and might not apply a transformation.

Downside is it assumes the DXIL version doesn't update UAddc for other types and if it ever did we would have to introduces some versioning to only do this for pre DXIL 1.6.x.

Upside it lets you keep the  direct mapping of `int_uadd_with_overflow` to `UAddc` and only modify for the expansion cases.

Reply:
All that said I don't think a solution here is pressing atm.

We could alternatively track this as technical debt we need to revisit as part of fully supporting `*_with_overflow` intrinsics. I wanted to use  these for DXIL ops documented here: https://github.com/llvm/llvm-project/issues/128638

That will give us time to develop a more complete solution.